File: Web/packages/kernel/test/webmcp-conformance.test.ts
//
//  webmcp-conformance.test.ts - the SHARED WebMCP corpus (OpenSource/Conformance/webmcp/
//  {project,registry}.json) through the TS fold. The Kotlin (WebMcpConformanceTest.kt) and
//  Swift (WebMcpConformance, via RecordMain.swift) twins run the SAME two files, because
//  the `<tool>` row is an authoring surface and the page table's law must be identical
//  wherever a shell implements it (proposals/webmcp.md).
//
//  Missing corpus = loud failure, never a silent zero-case pass.
//

import { test } from "node:test";
import assert from "node:assert/strict";
import { readFileSync, existsSync } from "node:fs";
import { join, dirname, resolve } from "node:path";

import {
  projectTools, webMcpResult, webMcpErrorResult, PageToolTable,
  type ToolRow, type PageToolRegistration, type PageToolRejection,
} from "../src/mcp/webmcp.ts";

function corpusFile(name: string): string {
  let dir = resolve(import.meta.dirname ?? ".");
  for (;;) {
    const candidate = join(dir, `OpenSource/Conformance/webmcp/${name}`);
    if (existsSync(candidate)) return candidate;
    const parent = dirname(dir);
    if (parent === dir) throw new Error(`webmcp corpus ${name} not found`);
    dir = parent;
  }
}

type ProjectCase = {
  name: string;
  actions?: { [k: string]: { inputs?: string[] } };
  tools?: ToolRow[];
  result?: { value?: unknown; thrown?: unknown; correlationId?: string };
  expect?: {
    descriptors?: unknown[];
    descriptorNames?: string[];
    result?: unknown;
  };
  expectError?: { code: string; names: string[] };
};

type RegistryStep = {
  register?: { surface: string; origin: string; tool: PageToolRegistration };
  commit?: { surface: string; origin: string };
  abort?: { surface: string; name: string };
};

type RegistryCase = {
  name: string;
  steps: RegistryStep[];
  expect?: {
    tools?: unknown[];
    toolNames?: string[];
    rejections?: unknown[];
    events?: Array<{ event: string; surface: string }>;
  };
};

const projectDoc = JSON.parse(readFileSync(corpusFile("project.json"), "utf-8")) as { cases: ProjectCase[] };
const registryDoc = JSON.parse(readFileSync(corpusFile("registry.json"), "utf-8")) as { cases: RegistryCase[] };
assert.ok(projectDoc.cases.length > 0, "webmcp project corpus is empty");
assert.ok(registryDoc.cases.length > 0, "webmcp registry corpus is empty");

for (const c of projectDoc.cases) {
  test(`webmcp-project/${c.name}`, () => {
    if (c.result !== undefined) {
      const actual = "thrown" in c.result && c.result.thrown !== undefined
        ? webMcpErrorResult(c.result.correlationId ?? "")
        : webMcpResult(c.result.value);
      assert.deepEqual(JSON.parse(JSON.stringify(actual)), c.expect?.result);
      return;
    }

    const inputs = new Map<string, readonly string[]>();
    for (const [name, decl] of Object.entries(c.actions ?? {})) inputs.set(name, decl.inputs ?? []);
    const { descriptors, errors } = projectTools(c.tools ?? [], inputs);

    if (c.expectError !== undefined) {
      assert.ok(errors.length > 0, `expected ${c.expectError.code}, got a clean projection`);
      assert.deepEqual(errors.map((e) => e.code), errors.map(() => c.expectError!.code));
      assert.deepEqual(errors.map((e) => e.name), c.expectError.names);
      // Every error carries a message that names the offending tool: a build failure whose
      // text does not say WHICH row is broken sends the author hunting.
      for (const e of errors) assert.ok(e.message.length > 0, "an error must carry a message");
      return;
    }

    assert.deepEqual(errors, [], `unexpected projection errors: ${JSON.stringify(errors)}`);
    if (c.expect?.descriptorNames !== undefined) {
      assert.deepEqual(descriptors.map((d) => d.name), c.expect.descriptorNames);
    }
    if (c.expect?.descriptors !== undefined) {
      assert.deepEqual(JSON.parse(JSON.stringify(descriptors)), c.expect.descriptors);
    }
  });
}

for (const c of registryDoc.cases) {
  test(`webmcp-registry/${c.name}`, () => {
    const events: Array<{ event: string; surface: string }> = [];
    const table = new PageToolTable((surface) => events.push({ event: "toolchange", surface }));
    const rejections: PageToolRejection[] = [];

    for (const step of c.steps) {
      if (step.register !== undefined) {
        const rejected = table.register(step.register.surface, step.register.origin, step.register.tool);
        if (rejected !== null) rejections.push(rejected);
      } else if (step.commit !== undefined) {
        table.commit(step.commit.surface, step.commit.origin);
      } else if (step.abort !== undefined) {
        table.abort(step.abort.surface, step.abort.name);
      } else {
        throw new Error(`unknown step: ${JSON.stringify(step)}`);
      }
    }

    if (c.expect?.tools !== undefined) {
      assert.deepEqual(JSON.parse(JSON.stringify(table.tools())), c.expect.tools);
    }
    if (c.expect?.toolNames !== undefined) {
      assert.deepEqual(table.tools().map((t) => t.name), c.expect.toolNames);
    }
    if (c.expect?.rejections !== undefined) {
      assert.deepEqual(JSON.parse(JSON.stringify(rejections)), c.expect.rejections);
    }
    if (c.expect?.events !== undefined) {
      assert.deepEqual(events, c.expect.events);
    }
  });
}

File: Engine/Android/core/src/test/kotlin/despia/engine/WebMcpConformanceTest.kt
package despia.engine

import org.junit.jupiter.api.DynamicTest
import org.junit.jupiter.api.TestFactory
import java.io.File
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * The WebMCP conformance runner — executes OpenSource/Conformance/webmcp/
 * {project,registry}.json through THIS runtime's WebMcp fold. The TS kernel
 * (webmcp-conformance.test.ts) and the Swift twin (WebMcpConformance) run the SAME two
 * files: the `<tool>` row is an authoring surface, so the unified-codebase law applies to
 * it in full, and the page table's law must be identical wherever a shell implements it.
 *
 * Missing corpus = loud failure, never a silent zero-case pass.
 */
class WebMcpConformanceTest {

    private fun corpusFile(name: String): File {
        var dir = File(System.getProperty("user.dir")).absoluteFile
        while (true) {
            val candidate = File(dir, "OpenSource/Conformance/webmcp/$name")
            if (candidate.isFile) return candidate
            dir = dir.parentFile ?: error("OpenSource/Conformance/webmcp/$name not found")
        }
    }

    private fun cases(name: String): List<Map<String, Any?>> {
        val doc = json(corpusFile(name).readText()).foundationValue as? Map<*, *>
            ?: error("$name: not a JSON object")
        @Suppress("UNCHECKED_CAST")
        val cases = doc["cases"] as? List<Map<String, Any?>> ?: error("$name: no cases[]")
        assertTrue(cases.isNotEmpty(), "$name is empty")
        return cases
    }

    /** Corpus JSON carries absent as NSNull; comparisons happen against real nulls. */
    private fun plain(v: Any?): Any? = when {
        v === NSNull -> null
        v is Map<*, *> -> v.entries.associate { (k, value) -> (k as String) to plain(value) }
        v is List<*> -> v.map { plain(it) }
        v is Int -> v.toDouble()
        v is Long -> v.toDouble()
        else -> v
    }

    @TestFactory
    fun projectCorpus(): List<DynamicTest> = cases("project.json").map { c ->
        DynamicTest.dynamicTest("webmcp-project/${c["name"]}") {
            @Suppress("UNCHECKED_CAST")
            val expect = plain(c["expect"]) as? Map<String, Any?> ?: emptyMap()

            // A RESULT case pins the MCP shaping a tool call answers with.
            @Suppress("UNCHECKED_CAST")
            val result = c["result"] as? Map<String, Any?>
            if (result != null) {
                val thrown = result["thrown"]
                val actual = if (thrown != null && thrown !== NSNull) {
                    WebMcpResult.error((plain(result["correlationId"]) as? String).orEmpty())
                } else {
                    WebMcpResult.value(plain(result["value"]))
                }
                assertEquals(plain(expect["result"]), plain(actual))
                return@dynamicTest
            }

            @Suppress("UNCHECKED_CAST")
            val actionsRaw = (plain(c["actions"]) as? Map<String, Any?>) ?: emptyMap()
            val actionInputs = LinkedHashMap<String, List<String>>()
            for ((name, decl) in actionsRaw) {
                @Suppress("UNCHECKED_CAST")
                val inputs = ((decl as? Map<String, Any?>)?.get("inputs") as? List<Any?>) ?: emptyList()
                actionInputs[name] = inputs.map { it as String }
            }

            @Suppress("UNCHECKED_CAST")
            val rows = ((plain(c["tools"]) as? List<Any?>) ?: emptyList()).map { raw ->
                val row = raw as Map<String, Any?>
                WebMcp.ToolRow(
                    action = row["action"] as? String ?: "",
                    description = row["description"] as? String ?: "",
                    asName = row["as"] as? String,
                    mutates = row["mutates"] as? String,
                )
            }

            val projection = WebMcp.project(rows, actionInputs)

            @Suppress("UNCHECKED_CAST")
            val expectError = plain(c["expectError"]) as? Map<String, Any?>
            if (expectError != null) {
                assertTrue(projection.errors.isNotEmpty(), "expected ${expectError["code"]}, got a clean projection")
                assertEquals(
                    List(projection.errors.size) { expectError["code"] },
                    projection.errors.map { it.code.wire },
                )
                assertEquals(expectError["names"], projection.errors.map { it.name })
                for (e in projection.errors) assertTrue(e.message.isNotEmpty(), "an error must carry a message")
                return@dynamicTest
            }

            assertEquals(emptyList(), projection.errors.map { it.message }, "unexpected projection errors")
            (expect["descriptorNames"] as? List<*>)?.let { names ->
                assertEquals(names, projection.descriptors.map { it.name })
            }
            (expect["descriptors"] as? List<*>)?.let { descriptors ->
                assertEquals(descriptors, projection.descriptors.map { plain(it.toMap()) })
            }
        }
    }

    @TestFactory
    fun registryCorpus(): List<DynamicTest> = cases("registry.json").map { c ->
        DynamicTest.dynamicTest("webmcp-registry/${c["name"]}") {
            val events = ArrayList<Map<String, Any?>>()
            val table = WebMcp.PageToolTable { surface ->
                events.add(linkedMapOf("event" to "toolchange", "surface" to surface))
            }
            val rejections = ArrayList<Map<String, Any?>>()

            @Suppress("UNCHECKED_CAST")
            val steps = (plain(c["steps"]) as? List<Any?>) ?: emptyList()
            for (raw in steps) {
                @Suppress("UNCHECKED_CAST")
                val step = raw as Map<String, Any?>
                @Suppress("UNCHECKED_CAST")
                val register = step["register"] as? Map<String, Any?>
                @Suppress("UNCHECKED_CAST")
                val commit = step["commit"] as? Map<String, Any?>
                @Suppress("UNCHECKED_CAST")
                val abort = step["abort"] as? Map<String, Any?>
                when {
                    register != null -> {
                        @Suppress("UNCHECKED_CAST")
                        val tool = register["tool"] as Map<String, Any?>
                        @Suppress("UNCHECKED_CAST")
                        val rejected = table.register(
                            surface = register["surface"] as String,
                            origin = register["origin"] as String,
                            name = tool["name"] as? String ?: "",
                            description = tool["description"] as? String ?: "",
                            inputSchema = tool["inputSchema"] as? Map<String, Any?>,
                            annotations = tool["annotations"] as? Map<String, Any?>,
                        )
                        if (rejected != null) {
                            rejections.add(linkedMapOf("reason" to rejected.reason.wire, "name" to rejected.name))
                        }
                    }
                    commit != null -> table.commit(commit["surface"] as String)
                    abort != null -> table.abort(abort["surface"] as String, abort["name"] as String)
                    else -> error("unknown step: $step")
                }
            }

            @Suppress("UNCHECKED_CAST")
            val expect = plain(c["expect"]) as? Map<String, Any?> ?: emptyMap()
            (expect["tools"] as? List<*>)?.let { tools ->
                assertEquals(tools, table.tools().map { wire(it) })
            }
            (expect["toolNames"] as? List<*>)?.let { names ->
                assertEquals(names, table.tools().map { it.name })
            }
            (expect["rejections"] as? List<*>)?.let { expected -> assertEquals(expected, rejections) }
            (expect["events"] as? List<*>)?.let { expected -> assertEquals(expected, events) }
        }
    }

    /** The recorded row as the corpus writes it — provenance, the verbatim schema, approval. */
    private fun wire(tool: WebMcp.PageTool): Map<String, Any?> {
        val out = LinkedHashMap<String, Any?>()
        out["surface"] = tool.surface
        out["origin"] = tool.origin
        out["name"] = tool.name
        out["description"] = tool.description
        out["inputSchema"] = plain(tool.inputSchema)
        if (tool.annotations != null) out["annotations"] = plain(tool.annotations)
        out["approval"] = tool.approval
        return out
    }
}
